website/source/docs/platform/k8s/connect.html.md

```diff
@@ -0,0 +1,251 @@
+to clients and dependencies exclusively over authorized and encrypted
+connections.
+
+-> **Note:** The pod specifications in this section are valid and use
+publicly available images. If you've installed the Connect injector, feel free
+to run the pod specifications in this section to try Connect with Kubernetes.
+
+### Accepting Inbound Connections
+
+An example pod is shown below with Connect enabled to accept inbound
+connections. Notice that the pod would still be fully functional without
+Connect. Minimal to zero modifications are required to pod specifications to
+enable Connect in Kubernetes.
+
+This pod specification starts an "echo" server that responds to any
+HTTP request with the static text "hello world".
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: echo-server
+  annotations:
+    "consul.hashicorp.com/connect-inject": "true"
+spec:
+  containers:
+    - name: echo-server
+      image: hashicorp/http-echo:latest
+      args:
+        - -text="hello world"
+        - -listen=:8080
+      ports:
+        - containerPort: 8080
+          name: http
+```
+
+The only change for Connect is the addition of the
+`consul.hashicorp.com/connect-inject` annotation. This enables injection
+for this pod. The injector can also be
+[configured](/docs/platform/k8s/connect.html#installation-and-configuration)
+to automatically inject unless explicitly disabled, but the default
+installation requires opt-in using the annotation shown above.
+
+This will start a Connect sidecar that listens on a random port registered
+with Consul and proxies valid inbound connections to port 8080 in the pod.
+To establish a connection to the pod, a client must use another Connect
+proxy. The client Connect proxy will use Consul service discovery to find
+all available upstream proxies and their public ports.
+
+In the example above, the server is listening on `:8080`. This means
+the server will still bind to the pod IP and allow external connections.
+This is useful to transition to Connect by allowing both Connect and
+non-Connect connections. To restrict only Connect connections, any listeners
+should bind to localhost only (such as `127.0.0.1`).
+
+### Connecting to Connect-Enabled Services
+
+The example pod specification below configures a pod that is capable
+of establishing connections to our previous example "echo" service. The
+connection to this echo service happens over an authorized and encrypted
+connection via Connect.
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: echo-client
+  annotations:
+    "consul.hashicorp.com/connect-inject": "true"
+    "consul.hashicorp.com/connect-service-upstreams": "echo-server:1234"
+spec:
+  containers:
+    - name: echo-client
+      image: tutum/curl:latest
+      # Just spin & wait forever, we'll use `kubectl exec` to demo
+      command: [ "/bin/sh", "-c", "--" ]
+      args: [ "while true; do sleep 30; done;" ]
```

Comment:
Hmm, would it be possible for our injector/admission controller (not quite sure the correct terminoligy now) to automatically populate ENV vars on the pod spec for each upstream?

That feels like a nice UX. In practice people may want to map them to their own ENV vars to not have to change the app but it makes it possible to just accept our ENV vars and make pod specs nicer, and even if not that, I _think_ you could still define the mapping in terms of the existing ENV var so that you only have to declare the upstream port in one place?

Long term we have other plans here so might not be worth the difference but if it's possible/easy that might be a nice touch for now?

Reply:
Done!